# --- T2-COPYRIGHT-BEGIN ---
# t2/package/*/firefox/hotfix.patch.x32
# Copyright (C) 2026 The T2 SDE Project
# SPDX-License-Identifier: GPL-2.0 or patched project license
# --- T2-COPYRIGHT-END ---

Start x32 by disabling x86 assembly, needs more bits e.g. xpcom, old patches:
  https://github.com/orsonteodoro/muslx32/commit/d3c69a81fc47b3b8a56c7ff13897e644eef6aedc#diff-4e39c3bf1b6aba655e5780d658e0af3005fe6a8b26b10f2c60ecc42f8a86edb3

--- firefox-119.0/python/mozbuild/mozbuild/configure/constants.py.vanilla	2023-10-31 18:39:29.027817314 +0100
+++ firefox-119.0/python/mozbuild/mozbuild/configure/constants.py	2023-10-31 18:40:10.919821943 +0100
@@ -65,7 +65,7 @@
     "sparc": 32,
     "sparc64": 64,
     "x86": 32,
-    "x86_64": 64,
+    "x86_64": 32,
     "wasm32": 32,
 }
 
--- firefox-142.0/js/src/vm/BigIntType.cpp.vanilla	2025-08-25 19:57:01.656453177 +0200
+++ firefox-142.0/js/src/vm/BigIntType.cpp	2025-08-25 19:58:22.066559403 +0200
@@ -317,7 +317,7 @@
 BigInt::Digit BigInt::digitDiv(Digit high, Digit low, Digit divisor,
                                Digit* remainder) {
   MOZ_ASSERT(high < divisor, "division must not overflow");
-#if defined(__x86_64__)
+#if defined(__x86_64__) && !defined(__ILP32__)
   Digit quotient;
   Digit rem;
   __asm__("divq  %[divisor]"
--- firefox-142.0/media/ffvpx/libavcodec/moz.build.vanilla	2025-08-25 19:06:52.760355939 +0200
+++ firefox-142.0/media/ffvpx/libavcodec/moz.build	2025-08-25 19:07:03.629378025 +0200
@@ -7,7 +7,7 @@
 # Due to duplicate file names, we compile libavutil/x86 in its own
 # moz.build file.
 if CONFIG['FFVPX_ASFLAGS']:
-    if CONFIG['TARGET_CPU'] == 'x86' or CONFIG['TARGET_CPU'] == 'x86_64':
+    if CONFIG['TARGET_CPU'] == 'x86':
         DIRS += ['x86']
 
 if CONFIG['TARGET_CPU'] == 'aarch64':
--- firefox-148.0/media/ffvpx/libavutil/moz.build.vanilla	2026-03-05 12:26:10.756978416 +0100
+++ firefox-148.0/media/ffvpx/libavutil/moz.build	2026-03-05 12:26:28.495351828 +0100
@@ -7,7 +7,7 @@
 # Due to duplicate file names, we compile libavutil/x86 in its own
 # moz.build file.
 if CONFIG['FFVPX_ASFLAGS']:
-    if CONFIG['TARGET_CPU'] == 'x86' or CONFIG['TARGET_CPU'] == 'x86_64':
+    if CONFIG['TARGET_CPU'] == 'x86':
         DIRS += ['x86']
 
 if CONFIG['TARGET_CPU'] == 'aarch64':
--- firefox-142.0/media/libaom/moz.build.vanilla	2025-08-25 19:04:44.235204785 +0200
+++ firefox-142.0/media/libaom/moz.build	2025-08-25 19:05:04.622350355 +0200
@@ -10,27 +10,7 @@
 include('sources.mozbuild')
 
 # Linux, Mac and Win share file lists for x86* but not configurations.
-if CONFIG['TARGET_CPU'] == 'x86_64':
-    EXPORTS.aom += files['X64_EXPORTS']
-    SOURCES += files['X64_SOURCES']
-    USE_NASM = True
-    if CONFIG['OS_TARGET'] == 'WINNT':
-        ASFLAGS += [ '-I%s/media/libaom/config/win/x64/' % TOPSRCDIR ]
-        LOCAL_INCLUDES += [ '/media/libaom/config/win/x64/' ]
-        EXPORTS.aom += [ 'config/win/x64/config/aom_config.h' ]
-        SOURCES += files['X64_WIN_SOURCES']
-        # This code is not included in our PGO profile, and pointlessly
-        # PGO-optimizing it slows down our builds a lot.
-        NO_PGO = True
-    elif CONFIG['OS_TARGET'] == 'Darwin':
-        ASFLAGS += [ '-I%s/media/libaom/config/mac/x64/' % TOPSRCDIR ]
-        LOCAL_INCLUDES += [ '/media/libaom/config/mac/x64/' ]
-        EXPORTS.aom += [ 'config/mac/x64/config/aom_config.h' ]
-    else: # Android, Linux, BSDs, etc.
-        ASFLAGS += [ '-I%s/media/libaom/config/linux/x64/' % TOPSRCDIR ]
-        LOCAL_INCLUDES += [ '/media/libaom/config/linux/x64/' ]
-        EXPORTS.aom += [ 'config/linux/x64/config/aom_config.h' ]
-elif CONFIG['TARGET_CPU'] == 'x86':
+if CONFIG['TARGET_CPU'] == 'x86':
     EXPORTS.aom += files['IA32_EXPORTS']
     SOURCES += files['IA32_SOURCES']
     USE_NASM = True
--- firefox-142.0/media/libvpx/moz.build.vanilla	2025-08-25 19:19:31.483423844 +0200
+++ firefox-142.0/media/libvpx/moz.build	2025-08-25 19:20:04.244414306 +0200
@@ -12,24 +12,7 @@
 if CONFIG['VPX_USE_NASM']:
     USE_NASM = True
 
-# Linux, Mac and Win share file lists for x86* but not configurations.
-if CONFIG['TARGET_CPU'] == 'x86_64':
-    if CONFIG['OS_TARGET'] == 'WINNT':
-        EXPORTS.vpx += files['WIN_X64_EXPORTS']
-        SOURCES += files['WIN_X64_SOURCES']
-        ASFLAGS += [ '-I%s/media/libvpx/config/win/x64/' % TOPSRCDIR ]
-        LOCAL_INCLUDES += [ '/media/libvpx/config/win/x64/' ]
-    elif CONFIG['OS_TARGET'] == 'Darwin':
-        EXPORTS.vpx += files['MAC_X64_EXPORTS']
-        SOURCES += files['MAC_X64_SOURCES']
-        ASFLAGS += [ '-I%s/media/libvpx/config/mac/x64/' % TOPSRCDIR ]
-        LOCAL_INCLUDES += [ '/media/libvpx/config/mac/x64/' ]
-    else: # Android, Linux, BSDs, etc.
-        EXPORTS.vpx += files['LINUX_X64_EXPORTS']
-        SOURCES += files['LINUX_X64_SOURCES']
-        ASFLAGS += [ '-I%s/media/libvpx/config/linux/x64/' % TOPSRCDIR ]
-        LOCAL_INCLUDES += [ '/media/libvpx/config/linux/x64/' ]
-elif CONFIG['TARGET_CPU'] == 'x86':
+if CONFIG['TARGET_CPU'] == 'x86':
     if CONFIG['OS_TARGET'] == 'WINNT':
         EXPORTS.vpx += files['WIN_IA32_EXPORTS']
         SOURCES += files['WIN_IA32_SOURCES']
--- firefox-142.0/gfx/ycbcr/moz.build.vanilla	2025-08-25 20:01:37.492621104 +0200
+++ firefox-142.0/gfx/ycbcr/moz.build	2025-08-25 20:02:07.827630641 +0200
@@ -17,22 +17,6 @@
     'yuv_row_table.cpp',
 ]
 
-if CONFIG['INTEL_ARCHITECTURE']:
-    # These files use MMX and SSE2 intrinsics, so they need special compile flags
-    # on some compilers.
-    SOURCES += ['yuv_convert_sse2.cpp']
-    SOURCES['yuv_convert_sse2.cpp'].flags += CONFIG['SSE2_FLAGS']
-
-    # MSVC doesn't support MMX when targeting AMD64.
-    if CONFIG['CC_TYPE'] == 'clang-cl':
-        if CONFIG['TARGET_CPU'] == 'x86':
-            SOURCES += [
-                'yuv_convert_mmx.cpp',
-            ]
-    else:
-        SOURCES += ['yuv_convert_mmx.cpp']
-        SOURCES['yuv_convert_mmx.cpp'].flags += CONFIG['MMX_FLAGS']
-
 if CONFIG['CC_TYPE'] == 'clang-cl':
     if CONFIG['TARGET_CPU'] == 'x86_64' or \
        (CONFIG['TARGET_CPU'] == 'x86' and CONFIG['CC_TYPE'] == 'clang-cl'):
--- firefox-142.0/gfx/ycbcr/yuv_row_posix.cpp.vanilla	2025-08-25 20:18:25.180248172 +0200
+++ firefox-142.0/gfx/ycbcr/yuv_row_posix.cpp	2025-08-25 20:18:44.527037844 +0200
@@ -9,7 +9,7 @@
 
 extern "C" {
 
-#if defined(ARCH_CPU_X86_64)
+#if 0
 
 // We don't need CPUID guards here, since x86-64 implies SSE2.
 
--- firefox-142.0/gfx/ycbcr/yuv_convert.cpp.vanilla	2025-08-25 20:25:50.280242482 +0200
+++ firefox-142.0/gfx/ycbcr/yuv_convert.cpp	2025-08-25 20:26:13.503383380 +0200
@@ -528,7 +528,7 @@
                                  dest_pixel, width, source_dx);
     } else {
 // Specialized scalers and rotation.
-#if defined(MOZILLA_MAY_SUPPORT_SSE) && defined(_MSC_VER) && defined(_M_IX86) && !defined(__clang__)
+#if 0
       if(mozilla::supports_sse()) {
         if (width == (source_width * 2)) {
           DoubleYUVToRGB32Row_SSE(y_ptr, u_ptr, v_ptr,
--- firefox-142.0/mfbt/RandomNum.cpp.vanilla	2025-08-25 18:45:56.692673510 +0200
+++ firefox-142.0/mfbt/RandomNum.cpp	2025-08-25 18:57:51.823014910 +0200
@@ -41,7 +41,11 @@
 // Older glibc versions don't define SYS_getrandom, so we define it here if
 // it's not available. See bug 995069.
 #  if defined(__x86_64__)
+#   if defined(__ILP32__)
+#    define GETRANDOM_NR (318 | 0x40000000)
+#   else
 #    define GETRANDOM_NR 318
+#   endif
 #  elif defined(__i386__)
 #    define GETRANDOM_NR 355
 #  elif defined(__aarch64__)
--- firefox-142.0/security/nss/coreconf/config.gypi.vanilla	2025-08-25 19:28:22.969412253 +0200
+++ firefox-142.0/security/nss/coreconf/config.gypi	2025-08-25 19:30:21.380684834 +0200
@@ -239,8 +239,6 @@
       }],
       [ 'target_arch=="x64"', {
         'defines': [
-          'NSS_X64',
-          'NSS_USE_64',
         ],
         # For Windows.
         'msvs_settings': {
--- firefox-142.0/ipc/chromium/src/base/linux_memfd_defs.h.vanilla	2025-08-25 18:55:28.219077248 +0200
+++ firefox-142.0/ipc/chromium/src/base/linux_memfd_defs.h	2025-08-25 18:58:04.797177661 +0200
@@ -14,7 +14,11 @@
 // and various related constants either.
 
 #if defined(__x86_64__)
+# if defined(__ILP32__)
+#  define MEMFD_CREATE_NR (319 | 0x40000000)
+# else
 #  define MEMFD_CREATE_NR 319
+# endif
 #elif defined(__i386__)
 #  define MEMFD_CREATE_NR 356
 #elif defined(__aarch64__)
--- firefox-142.0/mozglue/misc/SSE.h.vanilla	2025-08-25 20:28:32.808075667 +0200
+++ firefox-142.0/mozglue/misc/SSE.h	2025-08-25 20:29:25.887695241 +0200
@@ -94,7 +94,7 @@
  *
  */
 
-#if defined(__GNUC__) && (defined(__i386__) || defined(__x86_64__))
+#if defined(__GNUC__) && (defined(__i386__))
 
 #  ifdef __MMX__
 // It's ok to use MMX instructions based on the -march option (or
@@ -190,7 +190,7 @@
 
 #  define MOZILLA_SSE_HAVE_CPUID_DETECTION
 
-#  if defined(__x86_64__)
+#  if 0
 // MMX is always available on AMD64.
 #    define MOZILLA_PRESUME_MMX
 // SSE is always available on AMD64.
--- ./toolkit/moz.configure.orig	2026-07-15 22:45:00.000000000 +0200
+++ ./toolkit/moz.configure	2026-07-25 11:55:14.070428530 +0200
@@ -857,13 +857,13 @@
 
 @depends(target, "--with-system-av1", when=compile_environment)
 def dav1d_asm(target, system_av1):
-    if not system_av1 and target.cpu in ("arm", "aarch64", "x86", "x86_64"):
+    if not system_av1 and target.cpu in ("arm", "aarch64", "x86"):
         return True
 
 
 @depends(target, "--with-system-av1", when=compile_environment)
 def dav1d_nasm(target, system_av1):
-    if not system_av1 and target.cpu in ("x86", "x86_64"):
+    if not system_av1 and target.cpu in ("x86"):
         return namespace(version="2.14", what="AV1")
 
 
@@ -921,7 +921,7 @@
     if (
         target.kernel in ("WINNT", "Linux")
         and target.os != "Android"
-        and target.cpu in ("x86", "x86_64")
+        and target.cpu in ("x86")
     ):
         if wmf:
             return ("widevine", "wmfcdm")
@@ -1457,7 +1457,6 @@
         os_match = target.os in ("OSX",)
 
     cpu_match = target.cpu in (
-        "x86_64",
         "arm",
         "aarch64",
         "x86",
@@ -1635,7 +1634,7 @@
         return (
             target.kernel == "Linux"
             and host.kernel == "Linux"
-            and target.cpu in ("arm", "aarch64", "x86", "x86_64")
+            and target.cpu in ("arm", "aarch64", "x86")
         )
 
     option(
@@ -2158,7 +2157,6 @@
 def llvm_dlltool_flags(target):
     arch = {
         "x86": "i386",
-        "x86_64": "i386:x86-64",
         "aarch64": "arm64",
     }[target.cpu]
 
@@ -2316,7 +2314,7 @@
 
     @depends(target, when=in_tree_vpx)
     def vpx_nasm(target):
-        if target.cpu in ("x86", "x86_64"):
+        if target.cpu in ("x86"):
             if target.kernel == "WINNT":
                 # Version 2.03 is needed for automatic safeseh support.
                 return namespace(version="2.03", what="VPX")
@@ -2462,8 +2460,6 @@
             return in_tree_jpeg_mips64
         elif target.cpu == "x86":
             return ("-DPIC", "-DELF")
-        elif target.cpu == "x86_64":
-            return ("-D__x86_64__", "-DPIC", "-DELF")
 
     @depends(target, when=depends("--with-system-jpeg")(lambda x: not x))
     def jpeg_nasm(target):
@@ -2590,8 +2586,6 @@
         elif target.cpu == "aarch64" or target.cpu == "arm":
             flags = ["-DPIC", "-DELF"]
             use_nasm = False
-    elif target.cpu == "x86_64":
-        flags = ["-D__x86_64__", "-DPIC", "-DELF", "-Pconfig_unix64.asm"]
     else:
         audio_only = True
 
@@ -2712,7 +2706,7 @@
     # We only produce dxc builds for x86_64 and aarch64.
     @depends(target)
     def dxc_supports_target(target):
-        return target.cpu in ("x86_64", "aarch64")
+        return target.cpu in ("aarch64")
 
     @depends(
         target,
@@ -2727,7 +2721,6 @@
         if not windows_sdk_dir and not artifact:
             return namespace(enabled=False)
         arch = {
-            "x86_64": "x64",
             "aarch64": "arm64",
         }.get(target.cpu, target.cpu)
 
@@ -3409,7 +3402,7 @@
 @depends(target)
 def oxidized_breakpad(target):
     if target.kernel == "Linux":
-        return target.cpu in ("aarch64", "arm", "x86", "x86_64")
+        return target.cpu in ("aarch64", "arm", "x86")
     return False
 
 
--- firefox-153.0/tools/profiler/lul/LulMain.cpp.vanilla
+++ firefox-153.0/tools/profiler/lul/LulMain.cpp
@@ -1682,7 +1682,28 @@ static __attribute__((noinline)) bool Ge
   // Get hold of the current unwind-start registers.
   UnwindRegs startRegs;
   memset(&startRegs, 0, sizeof(startRegs));
-#if defined(GP_ARCH_amd64)
+#if defined(GP_ARCH_amd64) && defined(__ILP32__)
+  // x32 ABI: amd64 registers, but 32-bit pointers.
+  volatile uintptr_t block[3];
+  static_assert(sizeof(block) == 12);
+  __asm__ __volatile__(
+      "leaq 0(%%rip), %%r15"
+      "\n\t"
+      "movl %%r15d, 0(%0)"
+      "\n\t"
+      "movl %%esp, 4(%0)"
+      "\n\t"
+      "movl %%ebp, 8(%0)"
+      "\n"
+      :
+      : "r"(&block[0])
+      : "memory", "r15");
+  startRegs.xip = TaggedUWord(block[0]);
+  startRegs.xsp = TaggedUWord(block[1]);
+  startRegs.xbp = TaggedUWord(block[2]);
+  const uintptr_t REDZONE_SIZE = 128;
+  uintptr_t start = block[1] - REDZONE_SIZE;
+#elif defined(GP_ARCH_amd64)
   volatile uintptr_t block[3];
   static_assert(sizeof(block) == 24);
   __asm__ __volatile__(

NS_InvokeByIndex hardcodes the LP64 vtable layout: an 8 byte slot stride, an
8 byte load of the vtable pointer and an 8 byte indirect call through memory.
On x32 the vtable pointer and every slot are 4 bytes, so the 8 byte load picks
up the following object member as the high half of the vtable address and the
call faults. Segfault seen in every XPConnect call into native code.

--- firefox-153.0/xpcom/reflect/xptcall/md/unix/xptcinvoke_asm_x86_64_unix.S.vanilla
+++ firefox-153.0/xpcom/reflect/xptcall/md/unix/xptcinvoke_asm_x86_64_unix.S
@@ -96,9 +96,18 @@ SYM(NS_InvokeByIndex):
   add rsp, 112
 
 # load the function pointer and call
+#ifdef __ILP32__
+# On x32 the vtable pointer and every vtable slot are 4 bytes wide, and an
+# indirect call through memory would read 8, so load the target explicitly.
+  lea eax, [r13d * 4]
+  add eax, [rdi]
+  mov eax, [eax]
+  call rax
+#else
   lea eax, [r13d * 8]
   add rax, [rdi]
   call [rax]
+#endif
 
 # r12 and r13 were pushed relative to the old stack pointer which is now the
 # frame pointer.


InvokeCopyToStack() widens each XPCOM argument into the uint64_t gpregs[]
array that NS_InvokeByIndex later loads into rsi/rdx/rcx/r8/r9 with full
64 bit moves.  Pointers are widened with a plain (uint64_t) cast, which on
x32 is a cast from a 32 bit pointer to a wider integer: gcc emits movslq and
sign extends.  Any pointer at or above 0x80000000 therefore reaches the
callee as 0xffffffffxxxxxxxx and faults on first dereference.

x32 puts the stack near the top of the 4 GB address space (0xfff5c000) and
maps libxul plus the malloc arenas well above 2 GB, so both the indirect
case (&s->val, typically stack) and the T_CHAR_STR / T_INTERFACE default
case are hit in practice.  Firefox segfaults during startup in strlen(),
called from nsTSubstring::Assign(const char*, size_type(-1)) on a "utf-8"
charset string passed through xptcall.

Casting via uintptr_t makes the widening a zero extension, and also drops
the -Wpointer-to-int-cast warning gcc emits here.

--- firefox-153.0/xpcom/reflect/xptcall/md/unix/xptcinvoke_x86_64_unix.cpp.vanilla
+++ firefox-153.0/xpcom/reflect/xptcall/md/unix/xptcinvoke_x86_64_unix.cpp
@@ -26,7 +26,7 @@ InvokeCopyToStack(uint64_t * gpregs, dou
 
     for (uint32_t i = 0; i < paramCount; i++, s++) {
         if (s->IsIndirect())
-            value = (uint64_t) &s->val;
+            value = (uint64_t)(uintptr_t) &s->val;
         else {
             switch (s->type) {
             case nsXPTType::T_FLOAT:                                break;
@@ -42,7 +42,7 @@ InvokeCopyToStack(uint64_t * gpregs, dou
             case nsXPTType::T_BOOL:   value = s->val.b;             break;
             case nsXPTType::T_CHAR:   value = s->val.c;             break;
             case nsXPTType::T_WCHAR:  value = s->val.wc;            break;
-            default:                  value = (uint64_t) s->val.p;  break;
+            default:                  value = (uint64_t)(uintptr_t) s->val.p; break;
             }
         }
 


--- firefox-153.0/security/nss/lib/freebl/intel-gcm.S.vanilla	2026-07-26 16:16:48.227406417 +0200
+++ firefox-153.0/security/nss/lib/freebl/intel-gcm.S	2026-07-26 16:16:48.227816423 +0200
@@ -401,6 +401,7 @@ cdecl(intel_aes_gcmENC):
 .set Htbl, %rdx
 .set len, %rcx
 .set KS,%r9
+.set KS32,%r9d
 .set NR,%r10d
 
 .set Gctx, %rdx
@@ -490,7 +491,11 @@ cdecl(intel_aes_gcmENC):
 
     vmovdqu  288(Gctx), CTR
     vmovdqu  272(Gctx), T
+#if defined(__ILP32__)
+    movl 304(Gctx), KS32
+#else
     mov  304(Gctx), KS
+#endif
 # AESContext->Nr
     mov  244(KS), NR
 
@@ -992,6 +998,7 @@ cdecl(intel_aes_gcmDEC):
 .set Htbl, %rdx
 .set len, %rcx
 .set KS,%r9
+.set KS32,%r9d
 .set NR,%r10d
 
 .set Gctx, %rdx
@@ -1025,7 +1032,12 @@ cdecl(intel_aes_gcmDEC):
     andq    $-16, %rsp
     vmovdqu 288(Gctx), CTR
     vmovdqu 272(Gctx), T
+# platform_AES_GCMContext->aes_context is a 32-bit pointer under ILP32 (x32)
+#if defined(__ILP32__)
+    movl    304(Gctx), KS32
+#else
     mov     304(Gctx), KS
+#endif
 # AESContext->Nr
     mov     244(KS), NR
 
--- firefox-153.0/security/nss/lib/freebl/intel-gcm-wrap.c.vanilla	2026-07-26 16:16:48.228509799 +0200
+++ firefox-153.0/security/nss/lib/freebl/intel-gcm-wrap.c	2026-07-26 16:16:48.229091802 +0200
@@ -42,11 +42,15 @@ void intel_aes_gcmINIT(unsigned char Htb
                        unsigned char *KS,
                        int NR);
 
+/* The assembly consumes the lengths below as full 64-bit register values, so
+   they must be declared as such: under ILP32 (x32) "unsigned long" would only
+   define the low half of the argument register. */
+
 /* Produces the final GHASH value */
 void intel_aes_gcmTAG(unsigned char Htbl[16 * 16],
                       unsigned char *Tp,
-                      unsigned long Mlen,
-                      unsigned long Alen,
+                      PRUint64 Mlen,
+                      PRUint64 Alen,
                       unsigned char *X0,
                       unsigned char *TAG);
 
@@ -54,7 +58,7 @@ void intel_aes_gcmTAG(unsigned char Htbl
    Operates on whole blocks only. Partial blocks should be padded externally. */
 void intel_aes_gcmAAD(unsigned char Htbl[16 * 16],
                       unsigned char *AAD,
-                      unsigned long Alen,
+                      PRUint64 Alen,
                       unsigned char *Tp);
 
 /* Encrypts and hashes the Plaintext.
@@ -63,13 +67,13 @@ void intel_aes_gcmAAD(unsigned char Htbl
 void intel_aes_gcmENC(const unsigned char *PT,
                       unsigned char *CT,
                       void *Gctx,
-                      unsigned long len);
+                      PRUint64 len);
 
 /* Similar to ENC, but decrypts the Ciphertext. */
 void intel_aes_gcmDEC(const unsigned char *CT,
                       unsigned char *PT,
                       void *Gctx,
-                      unsigned long len);
+                      PRUint64 len);
 
 struct platform_AES_GCMContextStr {
     unsigned char Htbl[16 * AES_BLOCK_SIZE];


--- firefox-153.0/security/nss/lib/freebl/intel-aes.S.vanilla	2026-07-26 17:28:10.366800860 +0200
+++ firefox-153.0/security/nss/lib/freebl/intel-aes.S	2026-07-26 17:28:10.376819712 +0200
@@ -4,7 +4,13 @@
 
 	.text
 
+/* AESContext.iv follows the freeblCipherFunc worker pointer, so it starts four
+ * bytes earlier when that pointer is 32 bits wide (ILP32 / x32). */
+#ifdef __ILP32__
+#define IV_OFFSET 252
+#else
 #define IV_OFFSET 256
+#endif
 
 /*
  * Warning: the length values used in this module are "unsigned int"
@@ -536,8 +542,7 @@ cdecl(intel_aes_encrypt_cbc_128):
 	testl	%r9d, %r9d
 	je	2f
 
-//	leaq	IV_OFFSET(%rdi), %rdx
-	leaq	256(%rdi), %rdx
+	leaq	IV_OFFSET(%rdi), %rdx
 
 	movdqu	(%rdx), %xmm0
 	movdqu	(%rdi), %xmm2
@@ -597,8 +602,7 @@ cdecl(intel_aes_encrypt_cbc_128):
 	.globl cdecl(intel_aes_decrypt_cbc_128)
 	.balign	16
 cdecl(intel_aes_decrypt_cbc_128):
-//	leaq	IV_OFFSET(%rdi), %rdx
-	leaq	256(%rdi), %rdx
+	leaq	IV_OFFSET(%rdi), %rdx
 
 	movdqu	(%rdx), %xmm0   /* iv */
 	movdqu	(%rdi), %xmm2   /* first key block */
@@ -1358,8 +1362,7 @@ cdecl(intel_aes_encrypt_cbc_192):
 	testl	%r9d, %r9d
 	je	2f
 
-//	leaq	IV_OFFSET(%rdi), %rdx
-	leaq	256(%rdi), %rdx
+	leaq	IV_OFFSET(%rdi), %rdx
 
 	movdqu	(%rdx), %xmm0
 	movdqu	(%rdi), %xmm2
@@ -1423,8 +1426,7 @@ cdecl(intel_aes_encrypt_cbc_192):
 	.globl cdecl(intel_aes_decrypt_cbc_192)
 	.balign	16
 cdecl(intel_aes_decrypt_cbc_192):
-//	leaq	IV_OFFSET(%rdi), %rdx
-	leaq	256(%rdi), %rdx
+	leaq	IV_OFFSET(%rdi), %rdx
 
 	movdqu	(%rdx), %xmm0
 	movdqu	(%rdi), %xmm2
@@ -2264,8 +2266,7 @@ cdecl(intel_aes_encrypt_cbc_256):
 	testl	%r9d, %r9d
 	je	2f
 
-//	leaq	IV_OFFSET(%rdi), %rdx
-	leaq	256(%rdi), %rdx
+	leaq	IV_OFFSET(%rdi), %rdx
 
 	movdqu	(%rdx), %xmm0
 	movdqu	(%rdi), %xmm8
@@ -2334,8 +2335,7 @@ cdecl(intel_aes_encrypt_cbc_256):
 	.globl cdecl(intel_aes_decrypt_cbc_256)
 	.balign	16
 cdecl(intel_aes_decrypt_cbc_256):
-//	leaq	IV_OFFSET(%rdi), %rdx
-	leaq	256(%rdi), %rdx
+	leaq	IV_OFFSET(%rdi), %rdx
 
 	movdqu	(%rdx), %xmm0
 	movdqu	(%rdi), %xmm2


--- firefox-153.0/security/nss/lib/freebl/rijndael.c.vanilla	2026-07-26 17:28:10.377561489 +0200
+++ firefox-153.0/security/nss/lib/freebl/rijndael.c	2026-07-26 17:28:10.377975533 +0200
@@ -878,6 +878,13 @@ FREEBL_CIPHER_WRAP(platform_AES_GCMConte
 
 #if defined(USE_HW_AES)
 #if defined(NSS_X86_OR_X64)
+/* The cbc entry points in intel-aes.S reach AESContext.iv through a hardcoded
+ * IV_OFFSET, which shifts with the width of the worker pointer ahead of it. */
+#ifdef __ILP32__
+PR_STATIC_ASSERT(offsetof(AESContext, iv) == 252);
+#else
+PR_STATIC_ASSERT(offsetof(AESContext, iv) == 256);
+#endif
 FREEBL_CIPHER_WRAP(AESContext, intel_aes_encrypt_ecb_128);
 FREEBL_CIPHER_WRAP(AESContext, intel_aes_decrypt_ecb_128);
 FREEBL_CIPHER_WRAP(AESContext, intel_aes_encrypt_cbc_128);
--- firefox-153.0/third_party/zstd/lib/common/portability_macros.h.vanilla	2026-07-26 17:28:10.378326056 +0200
+++ firefox-153.0/third_party/zstd/lib/common/portability_macros.h	2026-07-26 17:28:10.378664587 +0200
@@ -139,10 +139,13 @@
  * - We are compiling for x86-64 and either:
  *   - DYNAMIC_BMI2 is enabled
  *   - BMI2 is supported at compile time
+ *
+ * The loop reads HUF_DecompressFastArgs with a hardcoded 8-byte stride, so it
+ * is unusable under ILP32 (x32), where the pointer members are 4 bytes wide.
  */
 #if !defined(ZSTD_DISABLE_ASM) &&                                 \
     ZSTD_ASM_SUPPORTED &&                                         \
-    defined(__x86_64__) &&                                        \
+    defined(__x86_64__) && !defined(__ILP32__) &&                 \
     (DYNAMIC_BMI2 || defined(__BMI2__))
 # define ZSTD_ENABLE_ASM_X86_64_BMI2 1
 #else
 #
Fix x32 content process crashes: the glibc struct pthread TID-cache fixup used
the LP64 tcbhead_t size, so MaybeUpdateGlibcTidCache() poked offset 712 instead
of 648 and every fork server child died on the CHECK_EQ (MOZ_CRASH) right away.
On x32 the pointers and longs of tcbhead_t are 4 bytes, shrinking it to 640.

Note SandboxLaunch.cpp compiles namespace_sandbox.inc, not the vendored
sandbox/linux/services/namespace_sandbox.cc, so the .inc needs the fix too.

--- firefox-153.0/security/sandbox/linux/launch/glibc_hack/sandbox/linux/services/namespace_sandbox.cc.vanilla	2026-07-15 22:09:52.000000000 +0200
+++ firefox-153.0/security/sandbox/linux/launch/glibc_hack/sandbox/linux/services/namespace_sandbox.cc	2026-07-26 18:05:35.291851577 +0200
@@ -90,12 +90,15 @@ void TerminationSignalHandler(int sig) {
 // https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/descr.h;hb=95a73392580761abc62fc9b1386d232cd55878e9#l121
 struct glibc_pthread {
   union {
-#  if defined(ARCH_CPU_X86_64)
+#  if defined(ARCH_CPU_X86_64) && !defined(__ILP32__)
     // On x86_64, sizeof(tcbhead_t) > sizeof(void*)*24.
     // https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/x86_64/nptl/tls.h;hb=95a73392580761abc62fc9b1386d232cd55878e9#l65
     // For all other architectures, sizeof(tcbhead_t) <= sizeof(void*)*24.
     // https://sourceware.org/git/?p=glibc.git&a=search&h=HEAD&st=grep&s=%7D+tcbhead_t
     char header[704];
+#  elif defined(ARCH_CPU_X86_64)
+    // On x32 the pointers and longs of tcbhead_t are 4 bytes, shrinking it to 640.
+    char header[640];
 #  endif
     void* padding[24];
   } header;


--- firefox-153.0/security/sandbox/linux/launch/glibc_hack/namespace_sandbox.inc.vanilla	2026-07-26 18:50:29.827214476 +0200
+++ firefox-153.0/security/sandbox/linux/launch/glibc_hack/namespace_sandbox.inc	2026-07-26 18:50:05.750884929 +0200
@@ -4,12 +4,15 @@
 // https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/descr.h;hb=95a73392580761abc62fc9b1386d232cd55878e9#l121
 struct glibc_pthread {
   union {
-#if defined(ARCH_CPU_X86_64)
+#if defined(ARCH_CPU_X86_64) && !defined(__ILP32__)
     // On x86_64, sizeof(tcbhead_t) > sizeof(void*)*24.
     // https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/x86_64/nptl/tls.h;hb=95a73392580761abc62fc9b1386d232cd55878e9#l65
     // For all other architectures, sizeof(tcbhead_t) <= sizeof(void*)*24.
     // https://sourceware.org/git/?p=glibc.git&a=search&h=HEAD&st=grep&s=%7D+tcbhead_t
     char header[704];
+#elif defined(ARCH_CPU_X86_64)
+    // On x32 the pointers and longs of tcbhead_t are 4 bytes, shrinking it to 640.
+    char header[640];
 #endif
     void* padding[24];
   } header;
